pmlc/dialect/op_lib/OpLibWrapperGen.cpp

```diff
@@ -24,28 +27,175 @@
 using llvm::raw_ostream;
 using llvm::Record;
 using llvm::RecordKeeper;
+using mlir::GenRegistration;
 using mlir::StringRef;
+using mlir::tblgen::EnumAttr;
 using mlir::tblgen::Operator;
 
-static const char* TBLGEN_DEBUG = std::getenv("TBLGEN_DEBUG");
+namespace pmlc {
+namespace dialect {
+namespace op {
+
+using namespace pmlc::dialect::op;  // NOLINT [build/namespaces]
+
+namespace tblgen {
+
+static inline std::map<StringRef, StringRef> getAttributes(const Operator& op) {
+  std::map<StringRef, StringRef> attributes_;
+  for (auto& namedAttr : op.getAttributes()) {
+    const auto& name = namedAttr.name;
+    auto mlir_type = namedAttr.attr.getReturnType();
+    const auto& is_enum = namedAttr.attr.isEnumAttr();
+    if (is_enum) {
+      mlir_type = namedAttr.attr.getAttrDefName();
+    }
+    attributes_.insert({name, mlir_type});
+  }
+  return attributes_;
+}
+
+static inline std::map<StringRef, StringRef> getOperands(const Operator& op) {
+  std::map<StringRef, StringRef> operands_;
+  for (int index = 0; index < op.getNumOperands(); index++) {
+    auto& namedOperand = op.getOperand(index);
+    const auto& name = namedOperand.name;
+    const auto& mlir_type = namedOperand.constraint.getDescription();
+    operands_.insert({name, mlir_type});
+  }
+  return operands_;
+}
+
+static inline StringRef getReturnType(const Operator& op) {
+  StringRef type;
+  int n_results = op.getNumResults();
+  if (n_results > 1) {
+    type = "MLIR_LIST";
+  } else if (n_results == 0) {
+    type = "MLIR_VOID";
+  } else {
+    auto& namedResult = op.getResult(0);
+    type = namedResult.constraint.getDescription();
+  }
+  return type;
+}
+
+struct OpInfo {
+  StringRef name_;
+  StringRef returnType_;
+  std::map<StringRef, StringRef> attributes_;
+  std::map<StringRef, StringRef> operands_;
```

Comment:
Use a `MapVector` so that you don't loose the order: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-mapvector-h